### PR TITLE
Add bracket argument to aunique

### LINKED ISSUE #2397: remove [] from %aunique
%aunique in a path will by default add [] around the disambiguator.  If a disambiguator is found on one album but is empty on the other, then the album name will end with [].

There may be a reason to have the brackets as default, but it seems to me a better option for the user to set any characters wanted around the disambiguator string in the path format itself.

I can do a PR, but I figured it would be easier just to post here and discuss if necessary.  Here's the line of code:

https://github.com/beetbox/beets/blob/958ad430fc4854b2589ec3a2726b7ee269e5322e/beets/library.py#L1511

### PR DESCRIPTION
**Pull request by diomekes**
A little background as to why I felt this was needed:

If I have two albums that are disambiguated by `albumdisambig`, and one album has that field empty, then it would have an empty `[]`. In this case, I would only want the brackets gone.

Also, if I do something like `[$year%aunique{}]`, then I would have two sets of brackets, with the inner set possibly empty.  Plus, in this case, if the inner brackets are removed, then there is still a white space after the year and before the closing bracket.

So this change does a few things:

- It adds a third argument to %aunique{} that should be two characters to be used as brackets.
- It removes the blank space along with any brackets created from an empty disambiguator.
- It adds brackets when the fallback `album_id` is used as the disambiguator (I decided to add this for consistency, but maybe it was left out originally for a reason I'm not aware of).

This will close #2397.

Finally, I think adding the brackets to the `album_id ` fallback messed up the test for `library.py`.  I fixed it, but I don't know if I did it correctly.

And I'm totally fine if this request is considered unnecessary for whatever reason, especially if some of my changes look "ugly".

### CODE REVIEW

File: beets/library.py [hunk 1]
```diff
@@ -1447,7 +1447,7 @@ def tmpl_time(s, fmt):
         cur_fmt = beets.config['time_format'].as_str()
         return time.strftime(fmt, time.strptime(s, cur_fmt))
 
-    def tmpl_aunique(self, keys=None, disam=None):
+    def tmpl_aunique(self, keys=None, disam=None, bracket=None):
```
Comment (sampsyo) [inline beets/library.py]: Could you please add a description of `bracket` to the docstring?

File: beets/library.py [hunk 2]
```diff
@@ -1467,8 +1467,10 @@ def tmpl_aunique(self, keys=None, disam=None):
 
         keys = keys or 'albumartist album'
         disam = disam or 'albumtype year label catalognum albumdisambig'
+        bracket = bracket or '[]'
         keys = keys.split()
         disam = disam.split()
+        bracket = None if bracket is ' ' else list(bracket)
```
Comment (sampsyo) [inline beets/library.py]: - Is there a reason the condition is `bracket is ' '` instead of using the empty string `''`?
- `is` might work sometimes for comparing strings, but `==` is the right operator to use here.

```py
>>> 'a' is 'a'
True
>>> 'a' * 100 is 'a' * 100
False
>>> 'a' * 100 == 'a' * 100
True
```

File: beets/library.py [hunk 3]
```diff
@@ -1502,13 +1504,20 @@ def tmpl_aunique(self, keys=None, disam=None):
...
+            res = u' {1}{0}{2}'.format(album.id, bracket[0] if bracket else
+                                       u'', bracket[1] if bracket else u'')
             self.lib._memotable[memokey] = res
             return res
 
         # Flatten disambiguation value into a string.
         disam_value = album.formatted(True).get(disambiguator)
-        res = u' [{0}]'.format(disam_value)
+        res = u' {1}{0}{2}'.format(disam_value, bracket[0] if bracket else u'',
+                                   bracket[1] if bracket else u'')
```
Comment (sampsyo) [inline beets/library.py]: Maybe this and the prior line with bracket-formatting would be more legible if the two brackets were assigned into variables?

```py
if bracket:
    bracket_l = bracket[0]
    bracket_r = bracket[1]
else:
    bracket_l = u''
    bracket_r = u''
```

File: beets/library.py [hunk 4]
```diff
@@ -1502,13 +1504,20 @@ def tmpl_aunique(self, keys=None, disam=None):
 
         else:
             # No disambiguator distinguished all fields.
-            res = u' {0}'.format(album.id)
+            res = u' {1}{0}{2}'.format(album.id, bracket[0] if bracket else
+                                       u'', bracket[1] if bracket else u'')
```
Comment (sampsyo) [inline beets/library.py]: This is fine—I don't have a strong opinion about whether the brackets should appear here, but you're certainly right that it will look more consistent.

File: beets/library.py [hunk 5]
```diff
@@ -1502,13 +1504,20 @@ def tmpl_aunique(self, keys=None, disam=None):
...
 
         # Flatten disambiguation value into a string.
         disam_value = album.formatted(True).get(disambiguator)
-        res = u' [{0}]'.format(disam_value)
+        res = u' {1}{0}{2}'.format(disam_value, bracket[0] if bracket else u'',
+                                   bracket[1] if bracket else u'')
+
+        # Remove space and/or brackets if disambiguation value is empty
+        if bracket and res == u' ' + ''.join(bracket) or res.isspace():
+            res = u''
```
Comment (sampsyo) [inline beets/library.py]: Hmm… is there a cleaner way we can check for an empty disambiguation string? Perhaps, if this appears earlier, we can just use `if disam_value:` for the check.

File: docs/reference/pathformat.rst [hunk 6]
```diff
@@ -112,21 +112,26 @@ will expand to "[2008]" for one album and "[2010]" for the other. The
...
+suffices, an arbitrary number is used to distinguish the two albums.
 
 The default identifiers are ``albumartist album`` and the default disambiguators
 are ``albumtype year label catalognum albumdisambig``. So you can get reasonable
 disambiguation behavior if you just use ``%aunique{}`` with no parameters in
 your path forms (as in the default path formats), but you can customize the
 disambiguation if, for example, you include the year by default in path formats.
 
+The default characters used as brackets are ``[]``. If a single blank space is
+used, then the disambiguator will not be surrounded by anything.
```
Comment (sampsyo) [inline docs/reference/pathformat.rst]: For full clarity, maybe something like this:

> The default characters used as brackets are `[]`. To change this, provide a third argument to the `%aunique` function consisting of two characters: the left and right brackets. Or, to turn off bracketing entirely, use a single blank space.

File: test/test_library.py [hunk 7]
```diff
@@ -713,8 +713,8 @@ def test_use_fallback_numbers_when_identical(self):
         album2.year = 2001
         album2.store()
 
-        self._assert_dest(b'/base/foo 1/the title', self.i1)
-        self._assert_dest(b'/base/foo 2/the title', self.i2)
+        self._assert_dest(b'/base/foo [1]/the title', self.i1)
+        self._assert_dest(b'/base/foo [2]/the title', self.i2)
```
Comment (sampsyo) [inline test/test_library.py]: That looks like the right change to me!

File: beets/library.py (hunk 2 above)
Comment (diomekes) [inline beets/library.py]: The `bracket == ' '` is to check for a white space in the third argument for `%aunique` in order to remove any bracketing.

**Comment by diomekes** [comment]
How do I add an empty field in the tests?  My efforts yielded `[0000]` for `year = None` and `[1]` for everything else.

File: beets/library.py (hunk 2 above)
Comment (sampsyo) [inline beets/library.py]: Right—I guess the question is whether the empty string might be a better "use nothing" indicator than a space.

File: test/test_library.py [hunk 8]
```diff
@@ -730,6 +730,16 @@ def test_unique_sanitized(self):
         self._setf(u'foo%aunique{albumartist album,albumtype}/$title')
         self._assert_dest(b'/base/foo [foo_bar]/the title', self.i1)
 
+    def test_drop_empty_disam_string(self):
+        album1 = self.lib.get_album(self.i1)
+        album1.year = None
+        album1.store()
+        self._assert_dest(b'/base/foo/the title', self.i1)
```
Comment (sampsyo) [inline test/test_library.py]: Hmm; I suppose I assumed you had run into this with some music in your library. Maybe you can reuse the data from there? You could, for example, try setting some arbitrary field `foo` to be empty on one album and to have a known value on another, and then set up the `%aunique` invocation to use that as the only (or first) disambiguator.

Comment (diomekes) [inline test/test_library.py]: Yeah, that is what I ran into with my library.  I think adding the second album fixed it.

File: beets/library.py [hunk 9]
```diff
@@ -1467,8 +1469,18 @@ def tmpl_aunique(self, keys=None, disam=None):
 
         keys = keys or 'albumartist album'
         disam = disam or 'albumtype year label catalognum albumdisambig'
+        bracket = bracket or '[]'
         keys = keys.split()
         disam = disam.split()
+        bracket = None if bracket == ' ' else list(bracket)
```
Comment (sampsyo) [inline beets/library.py]: To avoid crashes, maybe this should check that `bracket` has length 2.

File: docs/reference/pathformat.rst [hunk 10]
```diff
@@ -112,21 +112,28 @@ will expand to "[2008]" for one album and "[2010]" for the other. The
...
 The default identifiers are ``albumartist album`` and the default disambiguators
 are ``albumtype year label catalognum albumdisambig``. So you can get reasonable
 disambiguation behavior if you just use ``%aunique{}`` with no parameters in
 your path forms (as in the default path formats), but you can customize the
 disambiguation if, for example, you include the year by default in path formats.
 
+The default characters used as brackets are ``[]``. To change this, provide a
+third argument to the ``%aunique`` function consisting of two characters: the left
+and right brackets. Or, to turn off bracketing entirely, use a single blank
+space.
```
Comment (sampsyo) [inline docs/reference/pathformat.rst]: Great; this sounds good! But just to triple-check: is it necessary for the "null option" here to be a single space? That seems a little counterintuitive, but if there's a technical reason we can't use the empty string instead it's fine.

Comment (diomekes) [inline docs/reference/pathformat.rst]: Well, I set it up to default to `[]`.  I could change it to default to nothing, and that way we can get rid of the whole single space thing.  Is that what you mean?

Comment (sampsyo) [inline docs/reference/pathformat.rst]: Right—I meant that, to get no brackets, you would use `%aunique{foo,bar,}` instead of `%aunique{foo,bar, }`. To use the default square brackets, you would leave the comma off: `%aunique{foo,bar}`. In other words, the third argument to the function would be `''` for no brackets and `None` for the default.

Comment (diomekes) [inline docs/reference/pathformat.rst]: OK, I think I got it.  I was trying to distinguish `''` from `None` with `' '`, but I guess that's not necessary.

**Comment by sampsyo** [comment]
Fantastic; thank you for finishing this up! It's all merged. :sparkles:
